tests/server_test/mod.rs

```diff
@@ -0,0 +1,370 @@
+/// The webhook secret used to validate that the webhook events are coming
+/// from the expected source.
+const WEBHOOK_SECRET: &str = "secret";
+
+/// A context used for running a test.
+///
+/// This is used for interacting with the triagebot process and handling API requests.
+struct ServerTestCtx {
+    /// The triagebot process handle.
+    child: Child,
+    /// Stdout received from triagebot, used for debugging.
+    stdout: Arc<Mutex<Vec<u8>>>,
+    /// Stderr received from triagebot, used for debugging.
+    stderr: Arc<Mutex<Vec<u8>>>,
+    /// Directory for the temporary Postgres database.
+    db_dir: PathBuf,
+    /// The address for sending webhooks into the triagebot binary.
+    triagebot_addr: SocketAddr,
+    /// The handle to the mock server which simulates GitHub.
+    #[allow(dead_code)] // held for drop
+    server: HttpServerHandle,
+}
+
+/// The main entry point for a test.
+///
+/// Pass the name of the test as the first parameter.
+fn run_test(test_name: &str) {
+    crate::assert_single_record();
+    let activities = crate::load_activities("tests/server_test", test_name);
+    if !matches!(activities[0], Activity::Webhook { .. }) {
+        panic!("expected first activity to be a webhook event");
+    }
+    let ctx = build(activities);
+    // Wait for the server side to find a webhook. This will then send the
+    // webhook to the triagebot binary.
+    loop {
+        let activity = ctx
+            .server
+            .hook_recv
+            .recv_timeout(Duration::new(60, 0))
+            .unwrap();
+        match activity {
+            Activity::Webhook {
+                webhook_event,
+                payload,
+            } => {
+                eprintln!("sending webhook {webhook_event}");
+                let payload = serde_json::to_vec(&payload).unwrap();
+                ctx.send_webhook(&webhook_event, payload);
+            }
+            Activity::Error { message } => {
+                panic!("unexpected server error: {message}");
+            }
+            Activity::Finished => break,
+            a => panic!("unexpected activity {a:?}"),
+        }
+    }
+}
+
+fn build(activities: Vec<Activity>) -> ServerTestCtx {
+    // Set up a directory where this test can store all its stuff.
+    let tmp_dir = Path::new(env!("CARGO_TARGET_TMPDIR")).join("local");
+    let test_num = TEST_COUNTER.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    let test_dir = tmp_dir.join(format!("t{test_num}"));
+    if test_dir.exists() {
+        std::fs::remove_dir_all(&test_dir).unwrap();
+    }
+    std::fs::create_dir_all(&test_dir).unwrap();
+
+    let db_dir = test_dir.join("db");
+    setup_postgres(&db_dir);
+
+    let server = HttpServer::new(activities);
+    // TODO: This is a poor way to choose a TCP port, as it could already
+    // be in use by something else.
+    let triagebot_port = NEXT_TCP_PORT.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
```

Comment:
Does it test if the port is already taken?

Another approach could be like [Wiremock does](https://github.com/LukeMathWalker/wiremock-rs/blob/224f618d3a44460ee7d2017b236dc3317eb13eda/src/mock_server/pool.rs): allocate a pool of free ports at startup.

Reply:
I'm not sure I'm clear on the suggestion about setting up a pool. Each triagebot process is independent, and wouldn't be able to easily "take" a port from a pool (sending file descriptors over sockets exists, but not something I would want to bother with).

I'm not too worried about using up a lot of ports. I would only expect for there to ever be a few dozen tests, and there should be over 15,000 ports available starting at 50,000.

I added some code to pick a free port, though it may not be very reliable.